test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp

```diff
@@ -24,207 +24,266 @@
+               Iterator4 val_res, Size n)
+    {
     }
-    key_head[12] = 0;
-    val_head[12] = 0;
-
-    // call algorithm
-    auto new_policy = oneapi::dpl::execution::make_device_policy<class reduce_by_segment_1>(q);
-    auto res1 = oneapi::dpl::reduce_by_segment(new_policy, key_head, key_head + n, val_head, key_res_head, val_res_head);
-
-    // check values
-    n = std::distance(key_res_head, res1.first);
-    for (auto i = 0; i != n; ++i) {
-        if (i < 4) {
-            ASSERT_EQUAL(key_res_head[i], i+1);
-            ASSERT_EQUAL(val_res_head[i], i+1);
-        } else if (i == 4 || i == 6) {
-            ASSERT_EQUAL(key_res_head[i], 1);
-            ASSERT_EQUAL(val_res_head[i], 2);
-        } else if (i == 5 || i == 7) {
-            ASSERT_EQUAL(key_res_head[i], 3);
-            ASSERT_EQUAL(val_res_head[i], 6);
-        } else if (i == 8) {
-            ASSERT_EQUAL(key_res_head[i], 0);
-            ASSERT_EQUAL(val_res_head[i], 0);
-        } else {
-            std::cout << "fail: unexpected values in output range\n";
+};
+
+#if !TEST_DPCPP_BACKEND_PRESENT
+const int max_n = 100000;
+#endif
+
+template <typename T, typename TestName>
+void
+test4buffers()
+{
+#if TEST_DPCPP_BACKEND_PRESENT
+    const sycl::queue& queue = my_queue; // usm requires queue
+#    if _PSTL_SYCL_TEST_USM
+    {
+        // Allocate space for data using USM.
+        auto sycl_deleter = [queue](T* mem) { sycl::free(mem, queue.get_context()); };
+        ::std::unique_ptr<T, decltype(sycl_deleter)> key_head(
+            (T*)sycl::malloc_shared(sizeof(T) * max_n, queue.get_device(), queue.get_context()), sycl_deleter);
+        ::std::unique_ptr<T, decltype(sycl_deleter)> val_head(
+            (T*)sycl::malloc_shared(sizeof(T) * max_n, queue.get_device(), queue.get_context()), sycl_deleter);
+        ::std::unique_ptr<T, decltype(sycl_deleter)> key_res_head(
+            (T*)sycl::malloc_shared(sizeof(T) * max_n, queue.get_device(), queue.get_context()), sycl_deleter);
+        ::std::unique_ptr<T, decltype(sycl_deleter)> val_res_head(
+            (T*)sycl::malloc_shared(sizeof(T) * max_n, queue.get_device(), queue.get_context()), sycl_deleter);
+
+        T* keys = key_head.get();
+        T* vals = key_head.get();
+        T* key_res = key_res_head.get();
+        T* val_res = val_res_head.get();
+
+        for (size_t n = 1; n <= max_n; n = n <= 16 ? n + 1 : size_t(3.1415 * n))
+        {
+            invoke_on_all_hetero_policies<0>()(test_reduce_by_segment(), keys, keys + n, vals, key_res, val_res, n);
         }
     }
+#    endif
 
-    // call algorithm on single element range
-    key_res_head[0] = 9;
-    val_res_head[0] = 9;
-
-    auto new_policy2 = oneapi::dpl::execution::make_device_policy<class reduce_by_segment_2>(q);
-    auto res2 = oneapi::dpl::reduce_by_segment(new_policy2, key_head, key_head + 1, val_head, key_res_head, val_res_head);
+    // create buffers
+    sycl::buffer<uint64_t, 1> key_buf{sycl::range<1>(max_n)};
```

Comment:
Let's use parameterized type instead of `uint64_t`. The same below within `test4buffers`. Moreover, I expect that that keys and values may have different types.

Reply:
Agreed. `KeyT` and `ValT` should be used here instead. The types of the keys and values are allowed to differ according to the specification, so I'll extend the test case.

Reply:
The test uses different Key and Value types now across host and device executions.

Reply:
Done.